// @genkit-ai/core/lib/utils.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var utils_exports = {};
__export(utils_exports, {
  deepEqual: () => deepEqual,
  deleteUndefinedProps: () => deleteUndefinedProps,
  featureMetadataPrefix: () => featureMetadataPrefix,
  getCurrentEnv: () => getCurrentEnv,
  isDevEnv: () => isDevEnv,
  stripUndefinedProps: () => stripUndefinedProps
});
module.exports = __toCommonJS(utils_exports);
function deleteUndefinedProps(obj) {
  for (const prop in obj) {
    if (obj[prop] === void 0) {
      delete obj[prop];
    } else {
      if (typeof obj[prop] === "object") {
        deleteUndefinedProps(obj[prop]);
      }
    }
  }
}
function stripUndefinedProps(input) {
  if (input === void 0 || input === null || Array.isArray(input) || typeof input !== "object") {
    return input;
  }
  const out = {};
  for (const key in input) {
    if (input[key] !== void 0) {
      out[key] = stripUndefinedProps(input[key]);
    }
  }
  return out;
}
function getCurrentEnv() {
  return process.env.GENKIT_ENV || "prod";
}
function isDevEnv() {
  return getCurrentEnv() === "dev";
}
function featureMetadataPrefix(name) {
  return `feature:${name}`;
}
function deepEqual(a, b) {
  if (a === b) return true;
  if (a == null || b == null) return a === b;
  if (typeof a !== typeof b) return false;
  if (typeof a !== "object") return false;
  if (Array.isArray(a)) {
    if (!Array.isArray(b) || a.length !== b.length) return false;
    return a.every((v, i) => deepEqual(v, b[i]));
  }
  const aObj = a;
  const bObj = b;
  const aKeys = Object.keys(aObj).sort();
  const bKeys = Object.keys(bObj).sort();
  if (aKeys.length !== bKeys.length) return false;
  return aKeys.every(
    (key, i) => key === bKeys[i] && deepEqual(aObj[key], bObj[key])
  );
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  deepEqual,
  deleteUndefinedProps,
  featureMetadataPrefix,
  getCurrentEnv,
  isDevEnv,
  stripUndefinedProps
});
