// genkit/lib/beta.mjs
/**
 * @license
 *
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import {
  AgentInitSchema,
  AgentInputSchema,
  AgentOutputSchema,
  AgentStreamChunkSchema
} from "@genkit-ai/ai";
import { Session } from "@genkit-ai/ai/session";
import {
  AgentError,
  remoteAgent
} from "./client/agent.mjs";
import {
  InMemoryStreamManager,
  StreamNotFoundError
} from "@genkit-ai/core";
import { AsyncTaskQueue, lazy } from "@genkit-ai/core/async";
export * from "./common.mjs";
import {
  FileSessionStore,
  GenkitBeta,
  InMemorySessionStore,
  SessionRunner,
  applyPatch,
  diff,
  genkit
} from "./genkit-beta.mjs";
export {
  AgentError,
  AgentInitSchema,
  AgentInputSchema,
  AgentOutputSchema,
  AgentStreamChunkSchema,
  AsyncTaskQueue,
  FileSessionStore,
  GenkitBeta,
  InMemorySessionStore,
  InMemoryStreamManager,
  Session,
  SessionRunner,
  StreamNotFoundError,
  applyPatch,
  diff,
  genkit,
  lazy,
  remoteAgent
};
