// genkit/lib/beta.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __reExport = (target, mod, secondTarget) => (__copyProps(target, mod, "default"), secondTarget && __copyProps(secondTarget, mod, "default"));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var beta_exports = {};
__export(beta_exports, {
  AgentError: () => import_agent.AgentError,
  AgentInitSchema: () => import_ai.AgentInitSchema,
  AgentInputSchema: () => import_ai.AgentInputSchema,
  AgentOutputSchema: () => import_ai.AgentOutputSchema,
  AgentStreamChunkSchema: () => import_ai.AgentStreamChunkSchema,
  AsyncTaskQueue: () => import_async.AsyncTaskQueue,
  FileSessionStore: () => import_genkit_beta.FileSessionStore,
  GenkitBeta: () => import_genkit_beta.GenkitBeta,
  InMemorySessionStore: () => import_genkit_beta.InMemorySessionStore,
  InMemoryStreamManager: () => import_core.InMemoryStreamManager,
  Session: () => import_session.Session,
  SessionRunner: () => import_genkit_beta.SessionRunner,
  StreamNotFoundError: () => import_core.StreamNotFoundError,
  applyPatch: () => import_genkit_beta.applyPatch,
  diff: () => import_genkit_beta.diff,
  genkit: () => import_genkit_beta.genkit,
  lazy: () => import_async.lazy,
  remoteAgent: () => import_agent.remoteAgent
});
module.exports = __toCommonJS(beta_exports);
var import_ai = require("@genkit-ai/ai");
var import_session = require("@genkit-ai/ai/session");
var import_agent = require("./client/agent.js");
var import_core = require("@genkit-ai/core");
var import_async = require("@genkit-ai/core/async");
__reExport(beta_exports, require("./common.js"), module.exports);
var import_genkit_beta = require("./genkit-beta.js");
/**
 * @license
 *
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AgentError,
  AgentInitSchema,
  AgentInputSchema,
  AgentOutputSchema,
  AgentStreamChunkSchema,
  AsyncTaskQueue,
  FileSessionStore,
  GenkitBeta,
  InMemorySessionStore,
  InMemoryStreamManager,
  Session,
  SessionRunner,
  StreamNotFoundError,
  applyPatch,
  diff,
  genkit,
  lazy,
  remoteAgent,
  ...require("./common.js")
});
