// @genkit-ai/ai/lib/metadata.mjs
const GENKIT_UI_METADATA = {
  /**
   * Provides data to populate both standard and customized UI widgets.
   *
   * e.g.
   *
   * {
   *   [GENKIT_UI_METADATA.DATA_SOURCE]: {
   *     action: '/custom/foo',
   *     allowCustomValues: true
   *   }
   * }
   */
  DATA_SOURCE: "x-genkit-ui-data-source",
  /**
   * Provides a display name or label for the UI. By default, the Dev UI will
   * transform schema fields to "Sentence case". There are times where this is
   * not desirable and needs to be overwritten, such as abbreviations or
   * acronyms (e.g. "Top P", "Top K") and proper nouns (e.g. "Google Search
   * retrieval").
   *
   * e.g. { [GENKIT_UI_METADATA.DISPLAY_NAME]: 'Top P' }
   */
  DISPLAY_NAME: "x-genkit-ui-display-name",
  /**
   * Specifies the UI component (widget) to use for a schema field. Commonly
   * used for model and middleware configuration. Useful to resolve ambiguity
   * when multiple inputs could apply, or to provide a tailored user experience
   * for complex inputs.
   *
   * e.g. { [GENKIT_UI_METADATA.WIDGET]: GENKIT_UI_WIDGETS.MODEL_LIST }
   */
  WIDGET: "x-genkit-ui-widget"
};
const GENKIT_UI_WIDGETS = {
  /** A widget for configuring a list of models. */
  MODEL_LIST: "model-list-config",
  /** A widget for configuring LLM model safety settings. */
  SAFETY_SETTINGS: "safety-settings"
};
export {
  GENKIT_UI_METADATA,
  GENKIT_UI_WIDGETS
};
