// @genkit-ai/ai/lib/metadata.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var metadata_exports = {};
__export(metadata_exports, {
  GENKIT_UI_METADATA: () => GENKIT_UI_METADATA,
  GENKIT_UI_WIDGETS: () => GENKIT_UI_WIDGETS
});
module.exports = __toCommonJS(metadata_exports);
const GENKIT_UI_METADATA = {
  /**
   * Provides data to populate both standard and customized UI widgets.
   *
   * e.g.
   *
   * {
   *   [GENKIT_UI_METADATA.DATA_SOURCE]: {
   *     action: '/custom/foo',
   *     allowCustomValues: true
   *   }
   * }
   */
  DATA_SOURCE: "x-genkit-ui-data-source",
  /**
   * Provides a display name or label for the UI. By default, the Dev UI will
   * transform schema fields to "Sentence case". There are times where this is
   * not desirable and needs to be overwritten, such as abbreviations or
   * acronyms (e.g. "Top P", "Top K") and proper nouns (e.g. "Google Search
   * retrieval").
   *
   * e.g. { [GENKIT_UI_METADATA.DISPLAY_NAME]: 'Top P' }
   */
  DISPLAY_NAME: "x-genkit-ui-display-name",
  /**
   * Specifies the UI component (widget) to use for a schema field. Commonly
   * used for model and middleware configuration. Useful to resolve ambiguity
   * when multiple inputs could apply, or to provide a tailored user experience
   * for complex inputs.
   *
   * e.g. { [GENKIT_UI_METADATA.WIDGET]: GENKIT_UI_WIDGETS.MODEL_LIST }
   */
  WIDGET: "x-genkit-ui-widget"
};
const GENKIT_UI_WIDGETS = {
  /** A widget for configuring a list of models. */
  MODEL_LIST: "model-list-config",
  /** A widget for configuring LLM model safety settings. */
  SAFETY_SETTINGS: "safety-settings"
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  GENKIT_UI_METADATA,
  GENKIT_UI_WIDGETS
});
